File: src/minecraft/monnef/jaffas/xmas/ItemBlockXmasMulti.java
package monnef.jaffas.xmas;

import cpw.mods.fml.common.Side;
import cpw.mods.fml.common.asm.SideOnly;
import cpw.mods.fml.common.registry.LanguageRegistry;
import net.minecraft.src.ItemBlock;
import net.minecraft.src.ItemStack;
import net.minecraft.src.MathHelper;

public abstract class ItemBlockXmasMulti extends ItemBlock {
    private String[] subNames;
    private String[] subTitles;

    public ItemBlockXmasMulti(int id) {
        super(id);
        setHasSubtypes(true);
        this.subNames = this.getSubNames();
        this.subTitles = this.getSubTitles();
        this.registerNames(getParentBlock());
    }

    @Override
    public int getMetadata(int damageValue) {
        return damageValue;
    }

    @Override
    public String getItemNameIS(ItemStack itemstack) {
        return getItemName() + "." + subNames[itemstack.getItemDamage()];
    }

    public int getSubBlocksCount() {
        return this.subNames.length;
    }

    public void registerNames(BlockXmasMulti block) {
        for (int i = 0; i < subNames.length; i++) {
            ItemStack multiBlockStack = new ItemStack(block, 1, i);
            LanguageRegistry.addName(multiBlockStack, subTitles[multiBlockStack.getItemDamage()]);
        }
    }

    @SideOnly(Side.CLIENT)
    /**
     * Gets an icon index based on an item's damage value
     */
    public int getIconFromDamage(int par1) {
        int var2 = MathHelper.clamp_int(par1, 0, subNames.length);
        return this.iconIndex + var2;
    }

    protected abstract String[] getSubNames();

    protected abstract String[] getSubTitles();

    protected abstract BlockXmasMulti getParentBlock();
}

File: src/main/scala/monnef/jaffas/xmas/block/ItemBlockXmasMulti.scala
package monnef.jaffas.xmas.block

import monnef.jaffas.food.block.ItemBlockJaffas
import net.minecraft.item.ItemStack
import cpw.mods.fml.common.registry.LanguageRegistry
import cpw.mods.fml.relauncher.{SideOnly, Side}
import net.minecraft.util.{IIcon, MathHelper}
import monnef.core.common.CustomIconHelper
import monnef.core.api.ICustomIcon
import monnef.jaffas.xmas.common.IconDescriptorXmas
import net.minecraft.client.renderer.texture.IIconRegister
import net.minecraft.block.Block
import monnef.core.utils.RegistryUtils

abstract class ItemBlockXmasMulti(_block: Block) extends ItemBlockJaffas(_block) with IconDescriptorXmas {
  private var subTitles: Array[String] = null

  setHasSubtypes(true)
  this.subNames = this.getSubNames
  this.subTitles = this.getSubTitles

  override def getMetadata(damageValue: Int): Int = damageValue

  def getSubBlocksCount: Int = this.subNames.length

  @SideOnly(Side.CLIENT) override def getIconFromDamage(dmg: Int): IIcon = {
    val idx: Int = MathHelper.clamp_int(dmg, 0, subNames.length)
    icons(idx)
  }

  def getSubNames: Array[String]

  def getSubTitles: Array[String]

  def getParentBlock: BlockXmasMulti

  override def registerIcons(register: IIconRegister) {
    icons = (
      for (i <- 0 until getSubBlocksCount) yield register.registerIcon(CustomIconHelper.generateShiftedId(this.asInstanceOf[ICustomIcon], i))
      ).toArray
  }
}
